### ISSUE TITLE
Performance of SocketsHttpHandler when connected to an ONVIF Soap service

### ISSUE THREAD

**Issue by fabianeichfeldt** [dotnet/corefx#33703]
Hi,

I was sent from this issue [https://github.com/dotnet/wcf/issues/3272](https://github.com/dotnet/wcf/issues/3272).
@mconnew already supported me a lot and could isolate my problem to the new SocketsHttpHandler.

In short terms (please see the original issue for details):
I have a delay of round about 1s between each request to my SOAP service which is an IP based camera. 
After disabling the new SocketsHttpHandler with `AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);` this delay disappears.

I have created the requested tracing:
[net_000001.zip](https://github.com/dotnet/corefx/files/2618775/net_000001.zip)

Thanks in advance for your support.

**Comment by karelz** [comment]
Thanks for your report! Were you able to identify where the 1s delay is spent on the client?

**Comment by karelz** [comment]
cc @geoffkizer @wfurt

**Comment by fabianeichfeldt** [comment]
> Thanks for your report! Were you able to identify where the 1s delay is spent on the client

No, I couldn't spend the time yet to setup a debugable core framework

**Comment by freworr** [comment]
I was forwarded from the duplicate issue dotnet/core#2078. @Lxiamail Here is an update on the performance after disabling the SocketHttpHandler using the App switch.

````csharp
AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
````
The distribution of durations percentile shows the performance scores from within a 4 hour period.    
![old](https://user-images.githubusercontent.com/39900835/49366641-1e8ae780-f6ea-11e8-9508-58c4508bfc2c.PNG)
This shows that 99th percentile of the depedency requests duration scores 490 ms or better from a ASP .NET MV 5 app service. 
![new](https://user-images.githubusercontent.com/39900835/49366654-2480c880-f6ea-11e8-99b8-0aa4669073fe.PNG)
In comparison, the 99th percentile scores 9.8 s or better from a .NET Core app service. The distribution shows that there are some score outliers around 3 s and 10 s. Increasing the time period shows the same trends for both app services.

In conclusion, disabling the handler may have improved the dependency duration stability, but there are still some performance score outliers. The response times are still way faster and more stable in the ASP .NET MVC 5 app service.

How can we improve the performance to match the performance in the ASP .NET MVC 5?

**Comment by karelz** [comment]
@freworr I am confused which graph represents which run. Are they both .NET Core? First one is with SocketsHttpHandler, second one is with SocketsHttpHandler disabled?

If you're seeing perf gaps with SocketsHttpHandler disabled, then it would require separate investigation to analyze if it is something specific to your application, or if it can be reproduced with a simple standalone repro. We will prioritize it below/after investigating the perf gap of SocketsHttpHandler.

**Comment by freworr** [comment]
@karelz I've updated the comment to clarify which project the graphs correspond to.

**Comment by karelz** [comment]
Thanks! Does ASP.NET MVC 5 mean "on .NET Framework"?
If that's true, then my previous comment holds - we should focus first on fixing SocketsHttpHandler enabled. Once that is resolved we will likely not going to go after the perf problems when it is disabled.

**Comment by freworr** [comment]
Yes, it's on .NET Framework.

**Comment by freworr** [comment]
The distribution of dependency duration of enabled SocketsHttpHandler is similar to the disabled SocketsHttpHandler. It may have hit the connection timeout more often when it was enabled. So I can't see any significant improvement either.

**Comment by marquesoftware** [comment]
In some requests I have received the "CommunicationException" error after 21 seconds.

**Comment by freworr** [comment]
We have been testing the Httpclientfactory using named clients for a while now to consume SOAP requests moving away from the proxy class generated by adding a connected service. This seems to greatly improve the performance for the outbound requests. We did notice a strange behavior when the handler lifetime expired causing the outbound requests to timeout for a short period of time. So we increased the HTTP handler lifetime to 24 hours to minimize the frequency of timeout exceptions.

**Comment by freworr** [comment]
There seems to be an issue with the logical handler of the HttpClient looking at the received response from the same outbound request.
```txt
2019-03-29 13:39:52.673 +01:00 [Information] System.Net.Http.HttpClient.prognosisClient.ClientHandler: Received HTTP response after 166.0357ms - OK
2019-03-29 13:39:52.673 +01:00 [Information] System.Net.Http.HttpClient.prognosisClient.LogicalHandler: End processing HTTP request after 9724.0141ms - OK
``` 
The response is received relativly fast, but the logical handler use long time to process the request. A normal request response event againts the same external source would look like:
```txt
2019-03-29 13:39:52.977 +01:00 [Information] System.Net.Http.HttpClient.prognosisClient.ClientHandler: Received HTTP response after 136.0194ms - OK
2019-03-29 13:39:52.977 +01:00 [Information] System.Net.Http.HttpClient.prognosisClient.LogicalHandler: End processing HTTP request after 136.1589ms - OK
``` 
What is happening here?

**Comment by mconnew** [comment]
@freworr, is the .NET Core scenario running on Windows or Linux? If it's running on Windows, can you capture a profile using perfview and make sure to enable "Thread Time" data? I suspect the issue is caused by thread starvation as SocketsHttpHandler does different things with threads compared with HttpWebRequest. Also WCF has a few code paths where we make synchronous calls to code which then ends up async again so we end up using more thread pool threads. This can be easily verified by increasing workerThreads and completionPortThreads by calling ThreadPool.SetMinThreads. If things improve, then it verifies that's the problem and you have a workaround. Getting that perfview profile will then help with seeing where we can target fix any blocking thread usage.

**Comment by freworr** [comment]
We have been running the .NET Core application using an out-of-process hosting model with Kestrel in Azure, but after we adopted the in-process hosting model a few weeks ago, the issues we have experienced with the dependency calls have been solved.
```xml
<AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
<AspNetCoreModuleName>AspNetCoreModuleV2</AspNetCoreModuleName>
```
We were a bit late into the game of adopting this hosting model since we weren't aware of the possibility to do it in Azure (available from mid-January 2019?) before April.

**Comment by karelz** [comment]
OK, looks like some weird configuration problem, rather than product bug / inefficiency. Closing.

**Comment by fabianeichfeldt** [comment]
Just to propose my founded solution, for anyone who is struggling too:
The problem is the "Expect 100-continue" in the header which is enabled with new SocketHttpHandler. My requested SOAP servers doesn't handle these expectations properly. 
I think there is a timeout of 1000ms to wait for this continue telegram, isn't it?

In fact, with the postet solution of @mconnew [in wcf repository](https://github.com/dotnet/wcf/issues/3185) I could speed up my connections dramatically.
